relentless/core/data.py

```diff
@@ -0,0 +1,213 @@
+"""Core data management.
+
+Todo
+----
+Improve this documentation for developers!
+
+"""
+__all__ = ['Directory','Project']
+
+import os
+import shutil
+
+class Directory:
+    """Context for a filesystem directory.
+
+    The directory specified by `path` (which can be either absolute or relative)
+    is created if it does not already exist. This process is recursive, so
+    `path` may include multiple directories that do not yet exist. This object
+    represents the final directory in `path`.
+
+    A :py:class:`Directory` is a context that can be used to manage the current
+    working directory. Entering the context changes the current working
+    directory to *path*, and exiting restores the working directory before the
+    context was entered.
+
+    Parameters
+    ----------
+    path : str
+        Absolute or relative directory path.
+
+    Raises
+    ------
+    OSError
+        If the specified path is not a valid directory.
+
+    Examples
+    --------
+    Creating a directory::
+
+        d = Directory('foo')
+
+    Using the context to open a file `foo/bar.txt` in a directory::
+
+        with Directory('foo') as d:
+            f = open('bar.txt')
+
+    """
+    def __init__(self, path):
+        self.path = os.path.abspath(path)
+        if not os.path.exists(self.path):
+            os.makedirs(self.path)
+        self._start = []
+
+        if not os.path.isdir(self.path):
+            raise OSError('The specified path is not a valid directory.')
+
+    def __enter__(self):
+        """Enter the directory context.
+
+        The working directory is changed to the `path` of this object.
+
+        Returns
+        -------
+        :py:class:`Directory`
+            This directory.
+
+        """
+        self._start.append(os.getcwd())
+        os.chdir(self.path)
+
+        return self
+
+    def __exit__(self, exception_type, exception_value, traceback):
+        """Exit the directory context.
+
+        If possible, the working directory is reset to the path before entering
+        the context. The change is silently ignored if the original directory
+        no longer exists.
+
+        """
+        try:
+            if self._start[-1] is not os.getcwd():
+                os.chdir(self._start.pop())
```

Comment:
This will not pop the directory if you don't need to change, which means that the stack will get off by one.

I also ran a quick test on my system, and it costs 5x more to call `os.getcwd()` than it does to call `os.chdir()` to the cwd. So, I think there is no benefit to avoid calling `os.chdir`.

This should fix both:
```suggestion
            os.chdir(self._start.pop())
```